File: others/200-300/207. 课程表.py
import queue
from typing import List


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        deg = [0] * numCourses
        for x, y in prerequisites:
            graph[y].append(x)
            deg[x] += 1

        q = queue.Queue()
        cnt = 0
        for i in range(numCourses):
            if deg[i] == 0:
                cnt += 1
                q.put(i)

        while not q.empty():
            x = q.get()
            for v in graph[x]:
                deg[v] -= 1
                if deg[v] == 0:
                    q.put(v)
                    cnt += 1
        return cnt == numCourses


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        deg = [0] * numCourses
        for x, y in prerequisites:
            graph[y].append(x)
            deg[x] += 1

        def dfs(i: int) -> bool:
            if vis[i] == 1:
                return False

            if vis[i] == 2:
                return True

            vis[i] = 1
            res = True
            for v in graph[i]:
                res = res & dfs(v)

            vis[i] = 2
            return res

        ans = True
        vis = [0] * numCourses
        for i in range(numCourses):
            if vis[i] == 0:
                ans = ans & dfs(i)
                if not ans:
                    return False
        return ans

File: leetcode/200-300/207. 课程表.cpp
//
// Created by 123 on 25-5-26.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] =  {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

// dfs和拓扑

class Solution2 {
public:
  bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> graph(numCourses, vector<int>());
    for (auto& edge : prerequisites) {
      graph[edge[1]].push_back(edge[0]);
    }

    vector<int> visited(numCourses, 0);
    bool have_ring = false;
    auto dfs = [&](this auto&&dfs, int j) -> void {
      if (visited[j] == 1) {
        have_ring = true;
        return;
      }
      if (visited[j] == 2 || have_ring) {
        return;
      }
      visited[j] = 1;
      for (int neighbour : graph[j]) {
        dfs(neighbour);
      }
      visited[j] = 2;
    };
    for (int i = 0; i < numCourses; i++) {
      dfs(i);
      if (have_ring) {
        return !have_ring;
      }
    }
    return !have_ring;
  }
};

class Solution2 {
public:
  bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> graph(numCourses, vector<int>());
    vector<int> deg(numCourses, 0);
    for (auto& edge : prerequisites) {
      graph[edge[1]].push_back(edge[0]);
      deg[edge[0]]++;
    }

    queue<int> q;
    for (int i = 0; i < numCourses; i++) {
      if (deg[i] == 0)
        q.push(i);
    }

    int visited = 0;
    while (!q.empty()) {
      ++visited;
      int u = q.front();
      q.pop();
      for (int neighbour : graph[u]) {
        --deg[neighbour];
        if (deg[neighbour] == 0) {
          q.push(neighbour);
        }
      }
    }
    return visited == numCourses;
  }
};